parser/charset/encoding_base.go

```diff
@@ -0,0 +1,114 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package charset
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/pingcap/tidb/parser/mysql"
+	"github.com/pingcap/tidb/parser/terror"
+	"golang.org/x/text/encoding"
+	"golang.org/x/text/transform"
+)
+
+var errInvalidCharacterString = terror.ClassParser.NewStd(mysql.ErrInvalidCharacterString)
+
+// EncodingBase defines some generic functions.
+type EncodingBase struct {
+	enc encoding.Encoding
+}
+
+// Validate checks whether bytes are valid in current charset.
+func (b EncodingBase) Validate(src []byte) (nSrc int, ok bool) {
```

Comment:
A question here, not sure is this a good way.
Why we need `EncodingBase` type here? It has some overlap with `EncodingUTF8`. Maybe we can move the UTF8 implementation here?